### DISCUSSION TITLE
Chaining multiple python files

### DISCUSSION THREAD

**Discussion by Azuriye** [s0hv/foo_discord_rich#12] [Q&A] [answered]
Hello! Thanks for this awesome plugin to integrate foobar2k with Discord. In the Artwork upload command is it possible to chain multiple Python files? In this sense execute multiple scripts or maybe just extend the functionality of the current upload.py code?

Basically, I'm listening to radio stations and a series of two HTTP GET requests leads me to the URL where the image file is hosted, so I wanted to have some sort of condition where first it runs the normal upload.py code and if it fails to get any image, then jump to my custom code where it retrieves the image link hosted elsewhere... Is this possible? Secondly, since radio station tracks keep changing is it possible to call the upload.py multiple times upon each track change? because if it is, then I can think of ways to expand it to other radio stations which have album art...

**Comment by Azuriye** [comment]
So I did a bit more research and fb2k supports radio station album art through $info(cover_url) and it's even exposed in the fb2k API as now_playing_album_art_notify_manager, could you perhaps check if the current selected file which is playing is a radio stream or not?

JScript Panel has an example of utilizing these methods to update album art after every radio track change... Would be a nice to have feature...

**Comment by s0hv** [comment]
Seems like I didn't get an email notifications for this discussion. Anyways I'll look into this when I have the time. I haven't used radio stations with foobar2000 so if you have a url I can use to test it on that would help (along with any other required steps if any).

**Reply by Azuriye** [reply]
> Seems like I didn't get an email notifications for this discussion. Anyways I'll look into this when I have the time. I haven't used radio stations with foobar2000 so if you have a url I can use to test it on that would help (along with any other required steps if any).

Sorry for the late reply! You can get started over here...

https://hydrogenaud.io/index.php/topic,120855.html

A radio station you can try is 
https://radioparadise.com/listen/stream-links
It seems only the AAC streams contain the album art in them

The thread I've linked has several threads linked as well to it, mostly showing how album art for radio stations is being passed to foobar2k and how it handles it.

Though I wanted to propose something additional to it... Basically, I use JScript Panel to download my album art from a radio station which doesn't serve it through the URL but instead hosted somewhere else... and I save it as a .bin file along with my .mtags file and the JScript panels can correctly update the album art every time the radio station changes

Specifically using this function will check if the radio station's metadata or album art changes...
https://jscript-panel.github.io/docs/callbacks/#on_playback_dynamic_info_tracktype

So basically I want the plugin to keep checking if the radio station metadata (artist/track name) changes and an option to check if the album changes as well, sort of similar to that wiki I've linked... 

Here is how my files are ordered, a .mtag file and a .bin file which will only download if the track's metadata changes. Perhaps maybe the plugin could do the same and hopefully pick the .bin file every time it changes?
![image](https://github.com/s0hv/foo_discord_rich/assets/75412786/ef6ed804-a3e1-46b1-97b2-ec2c48a264e1)

Another Imgur post shows how to link both .mtags and have an album art for the radio station
https://imgur.com/a/4g1DPNG

Here it's static but in my case, as I keep downloading them the panels keep changing...

If you need any more info regarding this please let me know! will be happy to share additional info regarding this...

**Reply by s0hv** [reply]
Can you check out the [latest release](https://github.com/s0hv/foo_discord_rich/releases/tag/1.3.0-beta-5) and see if it works. I tested it with radio paradise and it worked fine. Your setup might not work correctly, as the plugin expects the local artwork to be changed the moment foobar2k changes the track title.

**Reply by Azuriye** [reply]
> Can you check out the [latest release](https://github.com/s0hv/foo_discord_rich/releases/tag/1.3.0-beta-5) and see if it works. I tested it with radio paradise and it worked fine. Your setup might not work correctly, as the plugin expects the local artwork to be changed the moment foobar2k changes the track title.

Hey! looks like it's working correctly with radio stations which support album art, and yeah you were right about how my setup wouldn't work correctly... could there be anything done regarding this? since not all radio stations serve their album art through ICE and instead host the images elsewhere on the internet... 

Perhaps some support for the JScript Panel plugin? That way when my panel script downloads the artwork and saves it, the panel could notify other panels and perhaps notify the discord RPC to change/refresh the artwork. currently, the plugin is reading the artwork properly (as a .bin file) and uploading it, but only the first image gets used so the album art on Discord is static instead of changing it from track to track.

In the JScript panel's documentation, this function is responsible for receiving notifications.
https://jscript-panel.github.io/docs/callbacks/#on_notify_dataname-info
Where this function is responsible for sending notifications to all panels.
https://jscript-panel.github.io/docs/namespaces/window/#windownotifyothersname-info

**Reply by s0hv** [reply]
I tried thinking of solutions but I don't really know how you could access stuff from this component in a JScript panel. As an image update is triggered on any track info change for streams, you could try modifying the metadata using [UpdateFileInfoFromJSON](https://jscript-panel.github.io/docs/interfaces/IMetadbHandleList/#updatefileinfofromjsonstr) after you've downloaded the cover art (like adding a single space). That might trigger the upload script again depending on what fb2k api it uses under the hood.

**Reply by Azuriye** [reply]
> I tried thinking of solutions but I don't really know how you could access stuff from this component in a JScript panel. As an image update is triggered on any track info change for streams, you could try modifying the metadata using [UpdateFileInfoFromJSON](https://jscript-panel.github.io/docs/interfaces/IMetadbHandleList/#updatefileinfofromjsonstr) after you've downloaded the cover art (like adding a single space). That might trigger the upload script again depending on what fb2k api it uses under the hood.

Just tested using that method... only issue is that when I write to the file the radio stream pauses and starts back basically entering into a loop where it will keep writing to the file... 

Perhaps if there is no other way of accessing JScript Panels, could the plugin add functionality to look for maybe a folder for changes? or a specific file extension/filename? I'm not sure how well this will work out or whether it'll pose some issues in the future but this sort of feels like the only way out...?

EDIT: I've just realised this now... forgot to mention but for some reason when I change the images (changing .bin file) the plugin doesn't seem to upload that album art but instead uses a very old album art unless I clear artwork URL and restart the playback for it to upload the new album art... Could it be because the .mTAG files haven't been updated hence using the old URL?

**Reply by s0hv** [reply]
I have two solutions that I think are viable. Solution 1 is a pre upload hook that runs another script before the artwork is extracted for upload. It'd be basically another command line command that would be run with the filename of the currently playing file as input. 

Solution 2 would be a toggleable option that disables the image cache only for streams without cover art urls. If the user decides to turn that option on, the upload script would always be run instead of using the cache. This way you could just wait for file changes in the upload script before returning a url (or just fetching and returning the url used for downloading) 

Solution 2 seems better to me as it's simpler to the user. 

Watching for file or folder changes in this component would be too specific of a feature to be useful.

**Reply by Azuriye** [reply]
Regarding Solution 2, suppose in my case the stream doesn't support album art through ICE if I enable that option I assume every time a new album art gets downloaded (since both RPC and FB2K can read the album art and display them properly without any additional hackery) it'll upload the album art through imgur and display it under RPC? it sounds very similar to watching for file/folder changes but not sure how it will all work out in the end 🤔

**Reply by s0hv** [reply]
> if I enable that option I assume every time a new album art gets downloaded it'll upload the album art through imgur and display it under RPC? 

Not exactly correct. The reason why it does not work currently, is because when the track changes the upload process begins. The component first reads the artwork data and computes an md5 hash. It then checks if any previously generated url matches that hash value and if it does it uses that url. Because this happens before the file is replaced by the latest cover, it'll find a matching url and use that, and in the case it does not find it the current image is uploaded before any other script has the time to update it. What solution 2 does is it would force the upload script to be run every time dynamic track changes happen. Then you could add whatever logic you need to the upload script itself to make sure the image is up to date.

There really is no clear solution to this other than quirky workarounds since I am not that familiar with the whole fb2k development environment and its features. Especially since this a pretty niche use case.

**Reply by s0hv** [reply]
Actually I thought of a really simple solution to this. Basically just adding a configurable wait time before trying to read and upload the cover art for streams without artwork data. Then you just need to set it high enough that the file is guaranteed to be updated. The drawback would be that the artwork would not be updated on discords side as fast as some other methods but it's so much more simple to implement.

Say the component waits for 5 seconds (or however long you configure it) and then runs all the things required for uploading. This would give time for your JScript panel to download the new album art and it would be reflected on discords side as well.

**Reply by Azuriye** [reply]
> What solution 2 does is it would force the upload script to be run every time dynamic track changes happen. Then you could add whatever logic you need to the upload script itself to make sure the image is up to date.

I feel like if the upload script could keep updating every time dynamic track changes happen it will solve it to some extent but another issue arises where if my panel downloads it way too late and the discord component already did its checks before the image has been downloaded, 

Perhaps Solution 2 + a configurable wait time which you mentioned as well would solve this issue up that way the user can narrowly increase or decrease the delay as they want... 

Hypothetically speaking downloading images isn't going to take minutes, at most maybe a minute?

**Reply by s0hv** [reply]
Downloading an image should take a few seconds at most on a decent connection unless it's a really high resolution image. 

I feel like it's a good tradeoff between much simpler code for a slightly less reliable way of making sure the image has changed. I'll leaves solution 2 out for the reason that it's more complicated to code and has its own quirks to account for. I'll update this thread when I've implemented the wait time solution.

**Reply by Azuriye** [reply]
Just a small doubt regarding 
> The component first reads the artwork data and computes an md5 hash. It then checks if any previously generated url matches that hash value and if it does it uses that url. Because this happens before the file is replaced by the latest cover, it'll find a matching url and use that, and in the case it does not find it the current image is uploaded before any other script has the time to update it.

Would implementing a delay effectively resolve the issue? Because currently let's say while fb2k is closed I change the album art of the radio stream... and open back fb2k and play the stream... for some reason, it's not uploading the newest artwork which I've replaced it with and instead uses the old artwork only unless I clear the discord URL which then sends the new album art...

Maybe the plugin isn't detecting any changes for radio streams when the album art changes locally? or maybe something along the lines of Artwork identifier key?

**Reply by s0hv** [reply]
Have you defined a duration for the stream somewhere in the metadata? Currently I check if it's a stream by checking if the duration is valid or not.

**Reply by Azuriye** [reply]
> Have you defined a duration for the stream somewhere in the metadata? Currently I check if it's a stream by checking if the duration is valid or not.

![image](https://github.com/s0hv/foo_discord_rich/assets/75412786/72bd3c21-dd1b-4775-8b40-090192fdb8b0)
Haven't mentioned it anywhere... should I mention some random value duration value?

**Reply by s0hv** [reply]
Can you check if console has any relevant logs regarding this specific situation

**Reply by Azuriye** [reply]
> Can you check if console has any relevant logs regarding this specific situation

```
Components loaded in: 0:00.068593
Configuration read in: 0:00.004862
foobar2000 v2.1 preview 2023-05-24 x86 [standard]
Watching: D:\Music\when musik
Playlist #0 loaded in 0:00.000103
JScript Panel 3.2.20 (Album Art by marc2003): initialised in 28 ms
Spider Monkey Panel v1.6.1 (Library Tree: Library Tree v2.2.0 by WilB): initialized in 9 ms
JScript Panel 3.2.20 (JS Playlist by Br3tt aka Falstaff): initialised in 28 ms
Spider Monkey Panel v1.6.1 (Biography: Biography v1.4.1 by WilB): initialized in 14 ms
User Interface initialized in: 0:00.664477
FFmpeg version: 6.0
Startup time : 0:00.806435
Library initialized after 0:00.828097
Opening track for playback: "D:\Music\when musik\Radio Stations\Gensokyo Radio.tags" / index: 1
m-TAGS:: Local prefixes: file:|unpack:|cdda:
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
Seekbar: taking Direct3D9 path.
Seekbar: Frontend initialized.
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
Automatic resampling: using Resampler (dBpoweramp/SSRC): 48000 Hz, Resampler (RetroArch): 48000 Hz
Device: Speakers (Realtek(R) Audio)
Mix format: 48000 Hz / 32-bit float / 2 channels (0x3)
Sending stream: 48000 Hz / 32-bit float / 2 channels (0x3)
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
Discord Rich Presence Integration v1.3.0-beta-5+638b551: connected to Azuriye
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
Biography Server: wikipedia: atabo: not found
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
m-TAGS:: REFERENCED FILE: https://stream.gensokyoradio.net/3 [0]
```

These are the logs from console when I'm doing this.
> Because currently let's say while fb2k is closed I change the album art of the radio stream... and open back fb2k and play the stream...

**Comment by marc2k3** [comment]
Can you not provide a menu item for uploads? Any scripting component could trigger it whenever they're ready to so without any logic on your side. You could hide it by default if it's too obscure for the average user.

**Reply by s0hv** [reply]
This is actually a good solution. No need for weird workarounds. A context menu command seems to be sufficient for JScript at least based on the `RunContextCommand` method.

**Comment by s0hv** [comment, marked as answer]
I added a "Regenerate artwork url" context menu option. Try to run the command with [RunContextCommand](https://jscript-panel.github.io/docs/interfaces/IMetadbHandleList/#runcontextcommandcommand) in your JScript panel after you've downloaded the new album art. When run successfully it should output stuff into console and update the url if the image changed.

**Reply by s0hv** [reply]
The context menu addition can be found in the latest release https://github.com/s0hv/foo_discord_rich/releases/tag/1.3.0-beta-6

**Reply by Azuriye** [reply]
> The context menu addition can be found in the latest release https://github.com/s0hv/foo_discord_rich/releases/tag/1.3.0-beta-6

Thanks a lot! I can confirm it's working now as intended with the Regenerate artwork url context menu 😊
